File: committee/committee.go
// Package committee is the Go consumer of the committee assignment that
// simctl/committee.py generates. It unmarshals committee.json into typed values and
// exposes per-node accessors; it holds no generation logic. The assignment is produced
// once, in Python (the topology seam: validators→nodes→committees→subnets), and passed
// in. The exported structs are exactly the committee.json schema, so the unmarshal
// output and a hand-built value are the same type — tests build arbitrary committees as
// in-memory literals, independent of the Python generator.
package committee

import (
	"encoding/json"
	"fmt"
	"os"
	"slices"
)

// Params are the assignment knobs (V, C, s_c independent; only C·s_c ≤ V is enforced,
// in the generator). Fields mirror committee.json.
type Params struct {
	N              int `json:"n"`
	V              int `json:"v"`
	C              int `json:"c"`
	Sc             int `json:"sc"`
	SubnetCount    int `json:"subnet_count"`
	SubnetsPerNode int `json:"subnets_per_node"`
	SubscribeFloor int `json:"subscribe_floor"`
	// TargetAggregators is the aggregators drawn per committee (clamped to the subnet's
	// subscriber count); each publishes one distinct aggregate.
	TargetAggregators int    `json:"target_aggregators"`
	Seed              uint64 `json:"seed"`
	NumSlots          int    `json:"num_slots"`
}

// AttesterRef is one committee seat: which node publishes, which validator, on which
// subnet, at which position within the committee.
type AttesterRef struct {
	Node     int `json:"node"`
	Val      int `json:"val"`
	Subnet   int `json:"subnet"`
	Position int `json:"position"`
}

// SlotPlan is one slot's committee draw (who attests where) plus its block proposer
// and aggregators.
type SlotPlan struct {
	Slot        int             `json:"slot"`
	Committees  [][]AttesterRef `json:"committees"`  // [committee] → its s_c attesters
	SubnetOf    []int           `json:"subnet_of"`   // [committee] → subnet id
	Proposer    int             `json:"proposer"`    // node that publishes this slot's block (a supernode)
	Aggregators [][]int         `json:"aggregators"` // [committee] → aggregator node ids
}

// Assignment is the whole run's plan: the stable per-subnet subscribe set plus the
// per-slot committee draws.
type Assignment struct {
	Params Params `json:"params"`
	// SubnetSubscribers[s] = nodes subscribing active subnet s (stable for the run); the
	// receivers/relayers a publisher dials into.
	SubnetSubscribers [][]int    `json:"subnet_subscribers"`
	Slots             []SlotPlan `json:"slots"`
}

// Load reads a committee.json produced by simctl/committee.py.
func Load(path string) (*Assignment, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	var a Assignment
	if err := json.Unmarshal(data, &a); err != nil {
		return nil, fmt.Errorf("parse %s: %w", path, err)
	}
	return &a, nil
}

// Subscribers returns the nodes subscribing subnet (stable for the run) — the expected
// receiver set for any attestation on it. Nil if subnet is out of range.
func (a *Assignment) Subscribers(subnet int) []int {
	if subnet < 0 || subnet >= len(a.SubnetSubscribers) {
		return nil
	}
	return a.SubnetSubscribers[subnet]
}

// ProposerSchedule returns the per-slot block proposer (a supernode), one entry per slot in
// slot order — the schedule the Validator obeys instead of the cyclic slot%N rule. Both
// backends read it from the same committee.json, so they propose identically.
func (a *Assignment) ProposerSchedule() []int {
	out := make([]int, len(a.Slots))
	for i, sp := range a.Slots {
		out[i] = sp.Proposer
	}
	return out
}

// AttestDuty is one attestation a node owes: which validator, on which subnet.
// Content-free — the vote is decided later, by the coupling.
type AttestDuty struct {
	Subnet int
	Val    int
}

// View narrows the assignment to one node's slice (what a single host runs).
type View struct {
	a    *Assignment
	node int
}

// Node returns the view of the assignment for one node.
func (a *Assignment) Node(node int) View { return View{a: a, node: node} }

// AttestDuties returns the attestations node owes this slot — one per hosted validator
// seated in one of the slot's committees (k>1 possible on the same subnet).
func (v View) AttestDuties(slot int) []AttestDuty {
	var duties []AttestDuty
	for _, com := range v.a.Slots[slot].Committees {
		for _, r := range com {
			if r.Node == v.node {
				duties = append(duties, AttestDuty{Subnet: r.Subnet, Val: r.Val})
			}
		}
	}
	return duties
}

// SubscribedSubnets returns the subnets this node subscribes (stable; it meshes on these
// for the whole run).
func (v View) SubscribedSubnets() []int {
	var out []int
	for subnet, members := range v.a.SubnetSubscribers {
		if slices.Contains(members, v.node) {
			out = append(out, subnet)
		}
	}
	return out
}

// AggregateSubnets returns the subnets this node aggregates this slot — the committees whose
// aggregator set includes it. It publishes one aggregate on each (on the global aggregate
// topic). Empty if the node isn't an aggregator this slot.
func (v View) AggregateSubnets(slot int) []int {
	if slot < 0 || slot >= len(v.a.Slots) {
		return nil
	}
	sp := v.a.Slots[slot]
	var out []int
	for ci, aggs := range sp.Aggregators {
		if slices.Contains(aggs, v.node) {
			out = append(out, sp.SubnetOf[ci])
		}
	}
	return out
}

File: simctl/committee.py
"""Committee assignment — the topology seam (validators→nodes→committees→subnets).

Pure, seeded source of truth for the attestation phase. Two parts:

- the per-slot **committee draw**: each slot, C committees of s_c validators (who attest,
  on which subnet); and
- the stable **subnet subscribe set**: for each active subnet, ≥``subscribe_floor`` nodes
  (the receivers/relayers), so every active subnet has a real mesh core. Publishers dial
  into this set per slot; see att-subnet.md.

`V`, `C`, `s_c` are independent knobs (only `C·s_c ≤ V`). The result serializes to
``committee.json`` (consumed by both backends + the analysis), so the assignment is
implemented once, here. Aggregation is out of scope for now.
"""

from __future__ import annotations

import random
from dataclasses import dataclass


@dataclass(frozen=True)
class Params:
    n: int  # nodes
    v: int  # validators
    c: int  # committees per slot (= active attestation subnets)
    sc: int  # committee size (attesters per committee per slot)
    subnet_count: int = 64
    subnets_per_node: int = 2  # subnets a node subscribes (capped at C)
    subscribe_floor: int = 10  # min subscribers per active subnet
    target_aggregators: int = 16  # aggregators per committee (TARGET_AGGREGATORS_PER_COMMITTEE)
    m: int = 1  # distinct aggregates each committee's aggregators publish
    seed: int = 1
    num_slots: int = 1


@dataclass(frozen=True)
class AttesterRef:
    node: int  # publishing node (gossip origin)
    val: int  # global validator index — the stable identity
    subnet: int  # committee → subnet
    position: int  # index within the committee


@dataclass
class SlotPlan:
    slot: int
    committees: list[list[AttesterRef]]  # [committee] → its s_c attesters
    subnet_of: list[int]  # [committee] → subnet id
    aggregators: list[list[int]]  # [committee] → aggregator node ids (subset of its subscribers)
    proposer: int = 0  # node that publishes this slot's block (a supernode; see generate)


@dataclass
class Assignment:
    params: Params
    subnet_subscribers: list[list[int]]  # [active subnet] → subscribing node ids (stable)
    slots: list[SlotPlan]

    def to_dict(self) -> dict:
        return {
            "params": {
                "n": self.params.n,
                "v": self.params.v,
                "c": self.params.c,
                "sc": self.params.sc,
                "subnet_count": self.params.subnet_count,
                "subnets_per_node": self.params.subnets_per_node,
                "subscribe_floor": self.params.subscribe_floor,
                "target_aggregators": self.params.target_aggregators,
                "m": self.params.m,
                "seed": self.params.seed,
                "num_slots": self.params.num_slots,
            },
            "subnet_subscribers": self.subnet_subscribers,
            "slots": [
                {
                    "slot": s.slot,
                    "committees": [[_ref_dict(r) for r in com] for com in s.committees],
                    "subnet_of": s.subnet_of,
                    "aggregators": s.aggregators,
                    "proposer": s.proposer,
                }
                for s in self.slots
            ],
        }


def _ref_dict(r: AttesterRef) -> dict:
    return {"node": r.node, "val": r.val, "subnet": r.subnet, "position": r.position}


def _rng(seed: int, *stream: int) -> random.Random:
    """A deterministic RNG keyed by integer scalars only — no str/tuple seeds, whose
    hashes are salted per-process (PYTHONHASHSEED) and would break reproducibility."""
    mixed = seed
    for s in stream:
        mixed = mixed * 1_000_003 + s
    return random.Random(mixed)


def params_from_v(v: int, n: int) -> Params:
    """Optional convenience: fill C, s_c from the mainnet formula
    ``C = min(64, V/4096)``, ``s_c = (V/32)/C`` for a realistic once-per-epoch point."""
    c = max(1, min(64, v // 4096))
    sc = (v // 32) // c
    return Params(n=n, v=v, c=c, sc=sc)


def generate(p: Params, supers: list[int] | None = None) -> Assignment:
    """Build the assignment. ``supers`` is the supernode id set (from
    topology.supernode_ids): block proposers are drawn from it round-robin over its sorted
    ids. Empty/None ⇒ cyclic over all nodes (``slot % n``), preserving the pre-supernode
    behavior for runs without supernodes."""
    if p.c * p.sc > p.v:
        raise ValueError(f"C*s_c ({p.c * p.sc}) > V ({p.v}): too many committee positions")
    if p.c > p.subnet_count:
        raise ValueError(f"C ({p.c}) > subnet_count ({p.subnet_count}): no committee→subnet map")
    subscribers = _subnet_subscribers(p)
    pool = sorted(supers) if supers else list(range(p.n))
    slots = [_slot_plan(p, slot, subscribers, pool[slot % len(pool)]) for slot in range(p.num_slots)]
    return Assignment(params=p, subnet_subscribers=subscribers, slots=slots)


def _subnet_subscribers(p: Params) -> list[list[int]]:
    """Stable subscribe set: every node subscribes ``min(subnets_per_node, C)`` random
    active subnets, then any active subnet below the floor is topped up — so each active
    subnet has ≥ ``min(subscribe_floor, N)`` subscribers (its mesh core)."""
    rng = _rng(p.seed, 1)
    spn = min(p.subnets_per_node, p.c)
    subs: list[set[int]] = [set() for _ in range(p.c)]
    for node in range(p.n):
        for s in rng.sample(range(p.c), spn):
            subs[s].add(node)
    floor = min(p.subscribe_floor, p.n)
    for s in range(p.c):
        if len(subs[s]) < floor:
            extra = [node for node in range(p.n) if node not in subs[s]]
            rng.shuffle(extra)
            subs[s].update(extra[: floor - len(subs[s])])
    return [sorted(s) for s in subs]


def _slot_plan(p: Params, slot: int, subscribers: list[list[int]], proposer: int) -> SlotPlan:
    # Independent per-slot draw: s_c·C distinct validators, chunked into C committees.
    vals = _rng(p.seed, 2, slot).sample(range(p.v), p.c * p.sc)
    agg_rng = _rng(p.seed, 3, slot)  # aggregator draw, independent of the committee draw
    committees: list[list[AttesterRef]] = []
    subnet_of: list[int] = []
    aggregators: list[list[int]] = []
    for ci in range(p.c):
        subnet = ci  # identity: committee ci → subnet ci (C ≤ subnet_count)
        members = [
            AttesterRef(node=v % p.n, val=v, subnet=subnet, position=pos)
            for pos, v in enumerate(vals[ci * p.sc : (ci + 1) * p.sc])
        ]
        committees.append(members)
        subnet_of.append(subnet)
        # Aggregators are drawn from the subnet's stable subscribers (they already receive
        # the attestations); ~target_aggregators of them, clamped to the subscriber count.
        subs = subscribers[subnet]
        k = min(p.target_aggregators, len(subs))
        aggregators.append(sorted(agg_rng.sample(subs, k)))
    return SlotPlan(slot=slot, committees=committees, subnet_of=subnet_of, aggregators=aggregators, proposer=proposer)
